### PR TITLE
fix(55014): Quick fix for ts7051 introduces incorrect type

### LINKED ISSUE #55014: Quick fix for ts7051 introduces incorrect type
# Bug Report

### 🔎 Search Terms

`quick fix`, `7051`, `ts7051`

### 🕗 Version & Regression Information

I am able to reproduce this behaviour with every version >= 4.5.5 available in the playground. There are no FAQ entries that mention `7051`.

### ⏯ Playground Link

1. [5.1.6](https://www.typescriptlang.org/play?ts=5.1.6#code/GYVwdgxgLglg9mABAEzgZTgWwKZQBYxgDmAFMAFyIkB0tYImARtgE4DaAugJSIC8AfInABrMHADuYHgG8AsAChEiFrhAskwEgEYANIgBMegMxcA3AoC+QA)
2. [5.2.0 beta](https://www.typescriptlang.org/play?ts=5.2.0-dev.20230713#code/GYVwdgxgLglg9mABAEzgZTgWwKZQBYxgDmAFMAFyIkB0tYImARtgE4DaAugJSIC8AfInABrMHADuYHgG8AsAChEiFrhAskwEgEYANIgBMegMxcA3AoC+QA)

### 💻 Code

```ts
function doSomething(f: (...number[]) => unknown) {
  //                     ~~~~~~~~~~
  // Parameter has a name but no type. Did you mean 'arg0: number[]'? (7051)
  return f(1, 2, 3);
}
```

Quick-fixed:

```ts
function doSomething(f: (...arg0: number[][]) => unknown) {
  return f(1, 2, 3);
  //       ~
  // Argument of type 'number' is not assignable to parameter of type 'number[]'. (2345)
}
```

### 🙁 Actual behavior

The suggestion is correct. However, `number[]` gets converted to `number[][]`.

### 🙂 Expected behavior

`(...number[]) => unknown` becomes `(...arg0: number[]) => unknown`.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #55014

### CODE REVIEW

File: src/services/codefixes/addNameToNamelessParameter.ts [hunk 1]
```diff
@@ -26,24 +33,43 @@ registerCodeFix({
...
-        param.dotDotDotToken ? factory.createArrayTypeNode(typeNode) : typeNode,
-        param.initializer);
-    changeTracker.replaceNode(sourceFile, param, replacement);
+        param.dotDotDotToken ? isArrayTypeNode(typeNode) ? typeNode : factory.createArrayTypeNode(typeNode) : typeNode,
+        param.initializer,
+    );
+    changeTracker.replaceRange(sourceFile, createRange(param.getStart(sourceFile), end), replacement);
+}
+
+function tryGetNextParam(sourceFile: SourceFile, param: ParameterDeclaration) {
```
Comment (DanielRosenwasser) [inline src/services/codefixes/addNameToNamelessParameter.ts:68]:
```suggestion
/**
 * If a user wrote
 *
 *     (string[][]) => string
 *
 * the parser will see this as an erroneously written
 *
 *     (string, [], []) => string
 *
 * Here we effectively reparse and return these binding pattern parameters
 * as long as we have subsequent '[]' pairs.
 */
 function tryGetNextArrayParam(sourceFile: SourceFile, param: ParameterDeclaration) {
```

File: src/services/codefixes/addNameToNamelessParameter.ts [hunk 2]
```diff
@@ -26,24 +33,43 @@ registerCodeFix({
...
+    const nextToken = findNextToken(param.name, param.parent, sourceFile);
+    if (nextToken && nextToken.kind === SyntaxKind.OpenBracketToken
+        && isArrayBindingPattern(nextToken.parent) && isParameter(nextToken.parent.parent)) {
```
Comment (DanielRosenwasser) [inline src/services/codefixes/addNameToNamelessParameter.ts:71]: I think you want to validate that there are no bound names inside of this thing.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: src/services/codefixes/addNameToNamelessParameter.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/services/codefixes/addNameToNamelessParameter.ts:71]: Could you provide a sample?

File: src/services/codefixes/addNameToNamelessParameter.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/services/codefixes/addNameToNamelessParameter.ts:68]: Do I understand correctly that this QF should handle only the following cases?

1. `(...string)` => `(...arg: string[])`
2. `(...string[])` =>  `(...arg: string[])`

Should the case with `...string[][]` be converted to `(...arg0: string[], [])` or `(...arg0: string[], arg1: string[])`?

Comment (sandersn) [reply src/services/codefixes/addNameToNamelessParameter.ts:68]: I think the code is correct as-is; I wouldn't limit it to 1 trailing `[]`.
(I didn't read @DanielRosenwasser 's comment as saying that it should be limited to 1.)

**Review by sandersn — CHANGES_REQUESTED** [review]

**Review by sandersn — APPROVED** [review]

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser could you review it?

File: src/services/codefixes/addNameToNamelessParameter.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/services/codefixes/addNameToNamelessParameter.ts:67]: Yeah, I'm just trying to clarify the logic with the comment

File: src/services/codefixes/addNameToNamelessParameter.ts (hunk 2 above)
Comment (DanielRosenwasser) [reply src/services/codefixes/addNameToNamelessParameter.ts:70]: Sure, something like

```ts
(Something[a]) => string
```

Comment (a-tarasyuk) [reply src/services/codefixes/addNameToNamelessParameter.ts:70]: @DanielRosenwasser This case doesn't trigger QF, `checker` handles only identifiers. We could also think about change that...

```ts
if (isIdentifier(param.name)) {
  const originalKeywordKind = identifierToKeywordKind(param.name);
  if (
    (isCallSignatureDeclaration(param.parent) ||
      isMethodSignature(param.parent) ||
      isFunctionTypeNode(param.parent)) &&
    param.parent.parameters.indexOf(param) > -1 &&
    (resolveName(
      param,
      param.name.escapedText,
      SymbolFlags.Type,
      /*nameNotFoundMessage*/ undefined,
      param.name.escapedText,
      /*isUse*/ true,
    ) ||
      (originalKeywordKind && isTypeNodeKind(originalKeywordKind)))
  ) {
    const newName = "arg" + param.parent.parameters.indexOf(param);
    const typeName =
      declarationNameToString(param.name) + (param.dotDotDotToken ? "[]" : "");
    errorOrSuggestion(
      noImplicitAny,
      declaration,
      Diagnostics.Parameter_has_a_name_but_no_type_Did_you_mean_0_Colon_1,
      newName,
      typeName,
    );
    return;
  }
}
```

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

File: src/services/codefixes/addNameToNamelessParameter.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/services/codefixes/addNameToNamelessParameter.ts]: @DanielRosenwasser Should I expand it to add support for `ElementAccessExpression`?

**Pull request merged**